components/movementsensor/replay/replay_utils_test.go

```diff
@@ -0,0 +1,337 @@
+	var dataIndex int
+	var err error
+	for i := 0; i < int(limit); i++ {
+		dataIndex, err = getNextDataAfterFilter(filter, last)
+		if err != nil {
+			if i == 0 {
+				return nil, err
+			}
+			continue
+		}
+
+		// Call desired function
+		data := createDataByMovementSensorMethod(filter.Method, dataIndex)
+
+		timeReq, timeRec, err := timestampsFromIndex(dataIndex)
+		if err != nil {
+			return nil, err
+		}
+
+		last = fmt.Sprint(dataIndex)
+
+		tabularData := &datapb.TabularData{
+			Data:          data,
+			TimeRequested: timeReq,
+			TimeReceived:  timeRec,
+		}
+		dataset = append(dataset, tabularData)
+	}
+
+	// Construct response
+	resp := &datapb.TabularDataByFilterResponse{
+		Data: dataset,
+		Last: last,
+	}
+
+	return resp, nil
+}
+
+// timestampsFromIndex uses the index of the data to generate a timeReceived and timeRequested for testing.
+func timestampsFromIndex(index int) (*timestamppb.Timestamp, *timestamppb.Timestamp, error) {
+	timeReq, err := time.Parse(time.RFC3339, fmt.Sprintf(testTime, index))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed parsing time")
+	}
+	timeRec := timeReq.Add(time.Second)
+	return timestamppb.New(timeReq), timestamppb.New(timeRec), nil
+}
+
+// getNextDataAfterFilter returns the index of the next data based on the provided.
+func getNextDataAfterFilter(filter *datapb.Filter, last string) (int, error) {
+	// Basic component part (source) filter
+	if filter.ComponentName != "" && filter.ComponentName != "source" {
+		return 0, ErrEndOfDataset
+	}
+
+	// Basic robot_id filter
+	if filter.RobotId != "" && filter.RobotId != validRobotID {
+		return 0, ErrEndOfDataset
+	}
+
+	// Basic location_id filter
+	if len(filter.LocationIds) == 0 {
+		return 0, errors.New("issue occurred with transmitting LocationIds to the cloud")
+	}
+	if filter.LocationIds[0] != "" && filter.LocationIds[0] != validLocationID {
+		return 0, ErrEndOfDataset
+	}
+
+	// Basic organization_id filter
+	if len(filter.OrganizationIds) == 0 {
+		return 0, errors.New("issue occurred with transmitting OrganizationIds to the cloud")
+	}
+	if filter.OrganizationIds[0] != "" && filter.OrganizationIds[0] != validOrganizationID {
+		return 0, ErrEndOfDataset
+	}
+
+	// Apply the time-based filter based on the seconds value in the start and end fields. Because our mock data
+	// does not have timestamps associated with them but are ordered we can approximate the filtering
+	// by sorting for the data in the list whose index is after the start second count and before the end second count.
+	// For example, if there are 15 entries the start time is 2000-01-01T12:00:10Z and the end time is 2000-01-01T12:00:14Z,
+	// we will return data from indices 10 to 14.
+	start := 0
+	end := defaultMaxDataLength[filter.Method]
+	if filter.Interval.Start != nil {
+		start = filter.Interval.Start.AsTime().Second()
+	}
+	if filter.Interval.End != nil {
+		end = int(math.Min(float64(filter.Interval.End.AsTime().Second()), float64(end)))
```

Comment:
instead of using math.Min here, can we check in `getData` if have returned the last data item in the data set and if so return ErrEndOfDataset

Reply:
thanks, this made it much easier to understand!